File: user_space/lib/flibc/execvp.zig
// flibc execvp — bare-name resolver over sys.exec_path.
//
// Linux's execvp consults $PATH; FlashOS has no environment yet, so
// a single search prefix is hard-wired: a bare name `foo` resolves to
// `/bin/foo`. A name that already contains a slash (absolute
// `/usr/bin/foo` or relative `tools/foo`) skips the prefix and is
// handed to `sys.exec_path` verbatim — the kernel joins relative paths
// against the task's `cwd` (slot 36).
//
// Layout mirrors readline.zig: the pure `resolve(name, out)` path-build
// lives at the top with host tests at the bottom of the file; the
// SVC-driving `execvp(name, argv)` driver sits behind an
// `if (has_driver)` struct-select so the host build never analyses the
// inline asm in syscalls.zig. The host fallback returns -1; only the
// aarch64-freestanding target sees the real `sys.exec_path` call.
//
// Path budget: PATH_MAX = 256 matches sys_chdir's CWD_SIZE and is the
// stack buffer fsh hands in. The resolver returns null on overflow
// (rule 1 — no realloc) so callers surface a clean -1 instead of
// truncating into a wrong binary.

const builtin = @import("builtin");

// Driver compiles only on aarch64-freestanding. Same gating idiom as
// readline.zig — the host-test build picks the empty branch so the
// SVC trampolines never reach semantic analysis.
const has_driver = builtin.cpu.arch == .aarch64 and builtin.target.os.tag == .freestanding;

/// Maximum resolved path length the driver hands to sys.exec_path. Sized
/// to match the kernel's `cwd` budget (CWD_SIZE = 256) — the kernel can
/// already handle paths up to that ceiling, so widening here would only
/// invite a later kernel-side rejection.
pub const PATH_MAX: usize = 256;

const BIN_PREFIX = "/bin/";

/// Resolve a program name into an absolute (or already-slashed) path
/// laid out in `out`. Returns a sentinel-terminated slice into `out`
/// suitable for `sys.exec_path`. Rules:
///   * empty `name`           → null
///   * `name` contains '/'    → copy verbatim + NUL into `out` (lets
///                              the kernel handle absolute / relative
///                              resolution against `cwd`)
///   * bare `name`            → `/bin/` + name + NUL
///   * `out` too small for    → null (caller gets -1 rather than a
///     prefix + name + NUL      silently truncated binary path)
///
/// Pure: no syscalls, no allocator. Exercised in isolation by the host
/// suite — see the `test` blocks at the bottom of this file.
pub fn resolve(name: []const u8, out: []u8) ?[:0]u8 {
    if (name.len == 0) return null;

    var has_slash = false;
    for (name) |c| {
        if (c == '/') {
            has_slash = true;
            break;
        }
    }

    if (has_slash) {
        if (name.len + 1 > out.len) return null;
        @memcpy(out[0..name.len], name);
        out[name.len] = 0;
        return out[0..name.len :0];
    }

    const total = BIN_PREFIX.len + name.len;
    if (total + 1 > out.len) return null;
    @memcpy(out[0..BIN_PREFIX.len], BIN_PREFIX);
    @memcpy(out[BIN_PREFIX.len..][0..name.len], name);
    out[total] = 0;
    return out[0..total :0];
}

/// Resolve `name` (bare → `/bin/<name>`; slashed → verbatim) and exec
/// the result. Returns -1 on resolve failure (empty / oversize) or
/// whatever `sys.exec_path` returns; on success the syscall does not
/// return.
pub const execvp = driver.execvp;

const driver = if (has_driver) struct {
    const sys = @import("syscalls.zig");

    pub fn execvp(name: [*:0]const u8, argv: [*]const ?[*:0]const u8) i32 {
        var path_buf: [PATH_MAX]u8 = undefined;
        var n: usize = 0;
        while (name[n] != 0) : (n += 1) {}
        const resolved = resolve(name[0..n], &path_buf) orelse return -1;
        return sys.exec_path(@ptrCast(resolved.ptr), argv);
    }
} else struct {
    // Host-test stub: never invoked from tests, present only so the
    // `pub const execvp = driver.execvp` binding type-checks on host.
    pub fn execvp(_: [*:0]const u8, _: [*]const ?[*:0]const u8) i32 {
        return -1;
    }
};

// ---- Host tests ----

const std = @import("std");
const testing = std.testing;

test "resolve: bare name maps to /bin/<name>" {
    var buf: [64]u8 = undefined;
    const r = resolve("fsh", &buf) orelse return error.UnexpectedNull;
    try testing.expectEqualStrings("/bin/fsh", r);
    try testing.expectEqual(@as(u8, 0), buf[r.len]);
}

test "resolve: single-char bare name" {
    var buf: [16]u8 = undefined;
    const r = resolve("x", &buf) orelse return error.UnexpectedNull;
    try testing.expectEqualStrings("/bin/x", r);
}

test "resolve: absolute path passes through verbatim" {
    var buf: [64]u8 = undefined;
    const r = resolve("/usr/local/bin/foo", &buf) orelse return error.UnexpectedNull;
    try testing.expectEqualStrings("/usr/local/bin/foo", r);
    try testing.expectEqual(@as(u8, 0), buf[r.len]);
}

test "resolve: relative path with slash passes through" {
    var buf: [64]u8 = undefined;
    const r = resolve("tools/foo", &buf) orelse return error.UnexpectedNull;
    try testing.expectEqualStrings("tools/foo", r);
}

test "resolve: leading '/' bypasses prefix even with no further slash" {
    var buf: [32]u8 = undefined;
    const r = resolve("/foo", &buf) orelse return error.UnexpectedNull;
    try testing.expectEqualStrings("/foo", r);
}

test "resolve: empty name returns null" {
    var buf: [64]u8 = undefined;
    try testing.expectEqual(@as(?[:0]u8, null), resolve("", &buf));
}

test "resolve: oversize bare name returns null" {
    var tiny: [4]u8 = undefined; // /bin/x = 6 chars + NUL = 7 > 4
    try testing.expectEqual(@as(?[:0]u8, null), resolve("foo", &tiny));
}

test "resolve: oversize passthrough returns null" {
    var tiny: [4]u8 = undefined;
    try testing.expectEqual(@as(?[:0]u8, null), resolve("/abcd", &tiny));
}

test "resolve: exact-fit bare name succeeds" {
    // "/bin/x" = 6 bytes, NUL = 7 → buffer of 7 fits
    var buf: [7]u8 = undefined;
    const r = resolve("x", &buf) orelse return error.UnexpectedNull;
    try testing.expectEqualStrings("/bin/x", r);
    try testing.expectEqual(@as(u8, 0), buf[6]);
}

test "resolve: one-byte-short bare name returns null" {
    var buf: [6]u8 = undefined; // /bin/x needs 7 with NUL
    try testing.expectEqual(@as(?[:0]u8, null), resolve("x", &buf));
}

test "resolve: exact-fit passthrough succeeds" {
    var buf: [5]u8 = undefined; // "/foo" = 4 + NUL = 5
    const r = resolve("/foo", &buf) orelse return error.UnexpectedNull;
    try testing.expectEqualStrings("/foo", r);
    try testing.expectEqual(@as(u8, 0), buf[4]);
}

test "resolve: bare name in PATH_MAX-sized buffer" {
    var buf: [PATH_MAX]u8 = undefined;
    const r = resolve("cat", &buf) orelse return error.UnexpectedNull;
    try testing.expectEqualStrings("/bin/cat", r);
}

test "resolve: slash mid-name treated as path" {
    var buf: [32]u8 = undefined;
    const r = resolve("a/b", &buf) orelse return error.UnexpectedNull;
    try testing.expectEqualStrings("a/b", r);
}

File: crates/flashsdk-base/src/execvp.rs
//! Bare-name program resolution on top of the exec syscall.
//!
//! Linux's `execvp` consults `$PATH`; FlashOS has no environment yet, so a single
//! search prefix is hard-wired: a bare name `foo` resolves to `/bin/foo`. A name that
//! already contains a slash -- absolute `/usr/bin/foo` or relative `tools/foo` -- skips
//! the prefix and is handed to the kernel verbatim, which joins relative paths against
//! the task's working directory.
//!
//! The path build in [`resolve`] is pure and host-tested; only [`execvp`], which traps
//! into the kernel, is target-gated.
//!
//! Path budget: [`PATH_MAX`] matches the kernel's working-directory ceiling and is the
//! stack buffer the shell hands in. The resolver reports overflow instead of
//! truncating, so a caller surfaces a clean `-1` rather than execing a wrong binary.

/// Maximum resolved path length handed to the exec syscall. Sized to match the
/// kernel's working-directory budget: the kernel already handles paths up to that
/// ceiling, so widening here would only invite a later kernel-side rejection.
pub const PATH_MAX: usize = 256;

const BIN_PREFIX: &[u8] = b"/bin/";

/// Resolve a program name into an absolute (or already-slashed) path laid out in
/// `out`. The rules are:
///
///   * an empty `name` yields `None`;
///   * a `name` containing `'/'` is copied verbatim, letting the kernel do the
///     absolute or cwd-relative resolution;
///   * a bare `name` becomes `/bin/` + `name`;
///   * an `out` too small for the result plus its terminator yields `None`, so the
///     caller gets `-1` rather than a silently truncated binary path.
///
/// On success the returned slice is the path bytes, and a NUL is written into `out`
/// one past its end: the returned slice's pointer is a valid C string for the exec
/// syscall. Pure -- no syscalls, no allocator.
pub fn resolve<'a>(name: &[u8], out: &'a mut [u8]) -> Option<&'a mut [u8]> {
    if name.is_empty() {
        return None;
    }

    let mut has_slash = false;
    for &c in name {
        if c == b'/' {
            has_slash = true;
            break;
        }
    }

    if has_slash {
        if name.len() + 1 > out.len() {
            return None;
        }
        out[..name.len()].copy_from_slice(name);
        out[name.len()] = 0;
        return Some(&mut out[..name.len()]);
    }

    let total = BIN_PREFIX.len() + name.len();
    if total + 1 > out.len() {
        return None;
    }
    out[..BIN_PREFIX.len()].copy_from_slice(BIN_PREFIX);
    out[BIN_PREFIX.len()..total].copy_from_slice(name);
    out[total] = 0;
    Some(&mut out[..total])
}

/// Resolve `name` (bare becomes `/bin/<name>`, slashed passes through) and exec the
/// result. Returns `-1` when resolution fails (empty or oversize name) or whatever the
/// exec syscall returns; on success the syscall does not return.
///
/// # Safety
///
/// `name` must point at a NUL-terminated string, and `argv` must be a NULL-terminated
/// vector of NUL-terminated pointers, all readable by the kernel for the length of the
/// call.
#[cfg(target_os = "none")]
pub unsafe fn execvp(name: *const u8, argv: *const *const u8) -> i32 {
    let mut path_buf = [0u8; PATH_MAX];
    let mut n: usize = 0;
    // SAFETY: the caller guarantees `name` is NUL-terminated, so the scan stops inside
    // the string's own allocation.
    while unsafe { *name.add(n) } != 0 {
        n += 1;
    }
    // SAFETY: `n` is the length measured above, and the bytes stay alive and unaliased
    // for the duration of this call.
    let bytes = unsafe { core::slice::from_raw_parts(name, n) };
    let resolved = match resolve(bytes, &mut path_buf) {
        Some(r) => r,
        None => return -1,
    };
    // SAFETY: `resolve` NUL-terminates the path one byte past `resolved`, and `argv`
    // carries the caller's terminator contract through unchanged.
    unsafe { crate::process::execve(resolved.as_ptr(), argv) }
}

#[cfg(test)]
mod tests {
    use super::*;

    /// The caller's buffer is uninitialized stack in the real call sites, so the
    /// tests hand in a poisoned one: a NUL assertion against a zeroed buffer would
    /// hold even if the resolver never wrote a terminator.
    const POISON: u8 = 0xaa;

    #[test]
    fn a_bare_name_maps_into_bin() {
        let mut buf = [POISON; 64];
        let r = resolve(b"fsh", &mut buf).expect("resolve returned None");
        assert_eq!(r, b"/bin/fsh");
        let len = r.len();
        assert_eq!(buf[len], 0);
    }

    #[test]
    fn a_single_char_bare_name_resolves() {
        let mut buf = [POISON; 16];
        let r = resolve(b"x", &mut buf).expect("resolve returned None");
        assert_eq!(r, b"/bin/x");
    }

    #[test]
    fn an_absolute_path_passes_through_verbatim() {
        let mut buf = [POISON; 64];
        let r = resolve(b"/usr/local/bin/foo", &mut buf).expect("resolve returned None");
        assert_eq!(r, b"/usr/local/bin/foo");
        let len = r.len();
        assert_eq!(buf[len], 0);
    }

    #[test]
    fn a_relative_path_with_a_slash_passes_through() {
        let mut buf = [POISON; 64];
        let r = resolve(b"tools/foo", &mut buf).expect("resolve returned None");
        assert_eq!(r, b"tools/foo");
    }

    #[test]
    fn a_leading_slash_bypasses_the_prefix_with_no_further_slash() {
        let mut buf = [POISON; 32];
        let r = resolve(b"/foo", &mut buf).expect("resolve returned None");
        assert_eq!(r, b"/foo");
    }

    #[test]
    fn an_empty_name_resolves_to_nothing() {
        let mut buf = [POISON; 64];
        assert!(resolve(b"", &mut buf).is_none());
    }

    #[test]
    fn an_oversize_bare_name_resolves_to_nothing() {
        // "/bin/foo" is 8 bytes plus a NUL: it cannot fit in 4.
        let mut tiny = [POISON; 4];
        assert!(resolve(b"foo", &mut tiny).is_none());
    }

    #[test]
    fn an_oversize_passthrough_resolves_to_nothing() {
        let mut tiny = [POISON; 4];
        assert!(resolve(b"/abcd", &mut tiny).is_none());
    }

    #[test]
    fn an_exact_fit_bare_name_succeeds() {
        // "/bin/x" is 6 bytes, and with the NUL that is 7: a 7-byte buffer fits.
        let mut buf = [POISON; 7];
        let r = resolve(b"x", &mut buf).expect("resolve returned None");
        assert_eq!(r, b"/bin/x");
        assert_eq!(buf[6], 0);
    }

    #[test]
    fn a_one_byte_short_bare_name_resolves_to_nothing() {
        // "/bin/x" needs 7 bytes once the NUL is counted.
        let mut buf = [POISON; 6];
        assert!(resolve(b"x", &mut buf).is_none());
    }

    #[test]
    fn an_exact_fit_passthrough_succeeds() {
        // "/foo" is 4 bytes plus the NUL: exactly 5.
        let mut buf = [POISON; 5];
        let r = resolve(b"/foo", &mut buf).expect("resolve returned None");
        assert_eq!(r, b"/foo");
        assert_eq!(buf[4], 0);
    }

    #[test]
    fn a_bare_name_resolves_in_a_path_max_buffer() {
        let mut buf = [POISON; PATH_MAX];
        let r = resolve(b"cat", &mut buf).expect("resolve returned None");
        assert_eq!(r, b"/bin/cat");
    }

    #[test]
    fn a_slash_mid_name_is_treated_as_a_path() {
        let mut buf = [POISON; 32];
        let r = resolve(b"a/b", &mut buf).expect("resolve returned None");
        assert_eq!(r, b"a/b");
    }
}
